client/beefy/src/worker.rs

```diff
@@ -212,116 +211,87 @@ where
-
-				// this metric is kind of 'fake'. Best BEEFY block should only be updated once we
-				// have a signed commitment for the block. Remove once the above TODO is done.
-				metric_set!(self, beefy_best_block, *notification.header.number());
-			}
+					.notify(|| Ok::<_, ()>(hash))
+					.expect("forwards closure result; the closure always returns Ok; qed.")
+			});
+			// Set new best BEEFY block number.
+			self.best_beefy_block = Some(block_num);
+			metric_set!(self, beefy_best_block, block_num);
+		} else {
+			debug!(target: "beefy", "🥩 Can't set best beefy to older: {}", block_num);
 		}
+	}
 
-		if self.should_vote_on(*notification.header.number()) {
-			let (validators, validator_set_id) = if let Some(rounds) = &self.rounds {
-				(rounds.validators(), rounds.validator_set_id())
-			} else {
-				debug!(target: "beefy", "🥩 Missing validator set - can't vote for: {:?}", notification.header.hash());
-				return
-			};
-			let authority_id = if let Some(id) = self.key_store.authority_id(validators) {
-				debug!(target: "beefy", "🥩 Local authority id: {:?}", id);
-				id
-			} else {
-				debug!(target: "beefy", "🥩 Missing validator id - can't vote for: {:?}", notification.header.hash());
-				return
-			};
-
-			let mmr_root =
-				if let Some(hash) = find_mmr_root_digest::<B, Public>(&notification.header) {
-					hash
-				} else {
-					warn!(target: "beefy", "🥩 No MMR root digest found for: {:?}", notification.header.hash());
-					return
-				};
-
-			let payload = Payload::new(known_payload_ids::MMR_ROOT_ID, mmr_root.encode());
-			let commitment = Commitment {
-				payload,
-				block_number: notification.header.number(),
-				validator_set_id,
-			};
-			let encoded_commitment = commitment.encode();
+	/// Handle session changes by starting new voting round for mandatory blocks.
+	fn init_session_at(&mut self, active: ValidatorSet<AuthorityId>, session_start: NumberFor<B>) {
+		debug!(target: "beefy", "🥩 New active validator set: {:?}", active);
+		metric_set!(self, beefy_validator_set_id, active.id());
+		// BEEFY should produce a signed commitment for each session
+		if active.id() != self.last_signed_id + 1 && active.id() != GENESIS_AUTHORITY_SET_ID {
+			metric_inc!(self, beefy_skipped_sessions);
+		}
 
-			let signature = match self.key_store.sign(&authority_id, &*encoded_commitment) {
-				Ok(sig) => sig,
-				Err(err) => {
-					warn!(target: "beefy", "🥩 Error signing commitment: {:?}", err);
-					return
-				},
-			};
+		if log_enabled!(target: "beefy", log::Level::Debug) {
+			// verify the new validator set - only do it if we're also logging the warning
+			let _ = self.verify_validator_set(&session_start, &active);
+		}
 
-			trace!(
-				target: "beefy",
-				"🥩 Produced signature using {:?}, is_valid: {:?}",
-				authority_id,
-				BeefyKeystore::verify(&authority_id, &signature, &*encoded_commitment)
-			);
+		let id = active.id();
+		self.rounds = Some(round::Rounds::new(session_start, active));
```

Comment:
For beefy mandatory block header, why we use new authorities set as `validator_set` for finalizing it. This looks to be a different behavior compared to grandpa. AFAKI, grandpa is using original authorities set as `validator_set` for provide finalize itself and provide justification proof for mandatory block.

We can tell that from `verify_justification ` order in the following implementation.

https://github.com/paritytech/parity-bridges-common/blob/a63d95ba7d1322477d2d941aacf3c47010d7aa4a/modules/grandpa/src/lib.rs#L161-L163

Reply:
That's a good point! From substrate-based-chain node point of view having the GRANDPA and BEEFY mandatory blocks overlap makes a lot of sense. As you suggest, [this commit](https://github.com/paritytech/substrate/pull/10882/commits/108519544b6e5767574b8c84bf5a34a5b27a8f8f) implements logic for the session boundary block (BEEFY Mandatory Block) to be signed by the ending-session's validator set.

Having said that, I want to point out that from a bridge point of view or light client point of view, this is rather irrelevant since light clients don't have to sync or care about headers, the [SignedCommitment](https://github.com/paritytech/substrate/blob/e75de5f0ed73e87f626d69d1479f3abee1aa6892/primitives/beefy/src/commitment.rs#L84) chain is what provides everything necessary.

BEEFY Payload should independently be able to provide consensus and proofs for the chain. In `polkadot` case, BEEFY payload is _Merkle root_ for [MmrLeaf](https://github.com/paritytech/substrate/blob/e75de5f0ed73e87f626d69d1479f3abee1aa6892/primitives/beefy/src/mmr.rs#L34) that also includes a field specifying [BeefyNextAuthoritySet](https://github.com/paritytech/substrate/blob/e75de5f0ed73e87f626d69d1479f3abee1aa6892/primitives/beefy/src/mmr.rs#L85).

As such the chain of trust can be verified using only `BEEFY SignedCommitment`s.